### PR TITLE
Properly detect which array and struct types are unsupported by the old ABI encoder.

### PR DESCRIPTION
**Pull request by chriseth**
Fixes https://github.com/ethereum/solidity/issues/3176

### CODE REVIEW

File: test/libsolidity/SolidityNameAndTypeResolution.cpp [hunk 1]
```diff
@@ -758,6 +758,62 @@ BOOST_AUTO_TEST_CASE(functions_with_stucts_of_non_external_types_in_interface_ne
...
+	)";
+	CHECK_WARNING(text, "Experimental features");
+}
+
+BOOST_AUTO_TEST_CASE(returning_arrays_in_structs_arrays)
+{
+	char const* text = R"(
+		contract C {
+			struct S { string[] s; }
+			function f() public pure returns (S x) {}
```
Comment (axic) [inline test/libsolidity/SolidityNameAndTypeResolution.cpp:1032]: Hm, why does this one needs a named return parameter while the other one doesn't?

Comment (chriseth) [reply test/libsolidity/SolidityNameAndTypeResolution.cpp:1032]: Probably a rebasing problem.

File: libsolidity/analysis/TypeChecker.cpp [hunk 2]
```diff
@@ -561,13 +584,12 @@ bool TypeChecker::visit(FunctionDefinition const& _function)
...
-			type(*var)->category() == Type::Category::Struct &&
-			!type(*var)->dataStoredIn(DataLocation::Storage) &&
-			!_function.sourceUnit().annotation().experimentalFeatures.count(ExperimentalFeature::ABIEncoderV2)
+			!_function.sourceUnit().annotation().experimentalFeatures.count(ExperimentalFeature::ABIEncoderV2) &&
+			!typeSupportedByOldABIEncoder(*type(*var))
 		)
 			m_errorReporter.typeError(
 				var->location(),
-				"Structs are only supported in the new experimental ABI encoder. "
+				"This type is only supported in the new experimental ABI encoder. "
```
Comment (axic) [inline libsolidity/analysis/TypeChecker.cpp:592]: Were the current (not the new ones) tests only checking for "only supported" or we didn't had tests to cover this?

Comment (chriseth) [reply libsolidity/analysis/TypeChecker.cpp:592]: We only had tests for structs, not structs inside arrays or arrays containing arrays etc.

File: libsolidity/ast/Types.cpp [hunk 3]
```diff
@@ -1578,8 +1578,6 @@ bool ArrayType::canBeUsedExternally(bool _inLibrary) const
 		return true;
 	else if (!m_baseType->canBeUsedExternally(_inLibrary))
 		return false;
-	else if (m_baseType->category() == Category::Array && m_baseType->isDynamicallySized())
-		return false;
```
Comment (axic) [inline libsolidity/ast/Types.cpp:1593]: So basically we enable multi level arrays now for non-experimental mode. Will it be still disabled due to no support for returndatacopy? Do we have a test for this?

Comment (chriseth) [reply libsolidity/ast/Types.cpp:1593]: It is still disabled by the type checker.

Comment (chriseth) [reply libsolidity/ast/Types.cpp:1593]: The function `typeSupportedByOldABIEncoder` is much stricter than what was there before.

**Comment by chriseth** [comment]
@axic anything still to be done here?

**Comment by chriseth** [comment]
Failure on travis is unrelated.

File: libsolidity/analysis/TypeChecker.cpp (hunk 2 above)
Comment (axic) [reply libsolidity/analysis/TypeChecker.cpp:592]: Yeah but none of the test expectations had to change so either there was no text or it didn't expect the entire message.

Comment (chriseth) [reply libsolidity/analysis/TypeChecker.cpp:592]: It seems we did not have tests for that.

Comment (chriseth) [reply libsolidity/analysis/TypeChecker.cpp:592]: Is this a blocker for merging? Please make clear statements.

Comment (axic) [reply libsolidity/analysis/TypeChecker.cpp:592]: Grepped the sources, there was no test for this before.

**Review by axic — APPROVED** [review]

File: libsolidity/analysis/TypeChecker.cpp (hunk 2 above)
Comment (axic) [reply libsolidity/analysis/TypeChecker.cpp:592]: No, there are tests now hopefully covering all of it.

**Pull request merged**